app/services/data_formatter.rb

```diff
@@ -0,0 +1,57 @@
+class DataFormatter
+  def initialize(purchases)
+    @purchases = purchases
+  end
+
+  def category_totals
+    Category.all.map do |c|
+      [c.name, sum_purchases_in_category(c)]
+    end
+  end
+
+  def category_totals_by_month
+    first_purchase = Purchase.all.order(:date).first[:date]
+    last_purchase = Purchase.all.order(:date).last[:date] # this could be Date.today rather than most recent purchase
+    months_in_range = (first_purchase..last_purchase).map { |d| d.strftime('%B %Y') }.uniq
+
+    sum_purchases_by_month(purchases_by_category, months_in_range)
+  end
+
+  private
+
+  def sum_purchases_in_category(category)
+    @purchases.where(subcategory: category.subcategories).to_a.map(&:amount).sum
+  end
+
+  def purchases_by_category
+    totals = Category.all.map do |c|
+      extract_purchases_by_category(c).map { |p| { name: c.name, date: p[:date], amount: p[:amount] } }
+    end
+
+    totals.flatten.group_by { |n| n[:name] }
+  end
+
+  def extract_purchases_by_category(category)
+    @purchases.where(subcategory: category.subcategories).map { |p| { date: p.date, amount: p.amount } }
```

Comment:
Actually, I'd say that this could now be a var inside the loop in `purchases_by_category`, rather than a separate method.

Also,in `purchases_by_category` when you loop over the purchases per category, you already have the `date` and `amount` k-v pairs, so you can probably just add in `name` rather than repeat the other two.

Reply:
Seems reasonable if you think that's not putting too much logic in the former.

And good suggestion - wasn't sure if it could be done, but .merge! seems to be the solution.